### PR TITLE
forcing floating point arithemtic for binomial_coefficient_log

### PR DESCRIPTION
**Pull request by syclik**
#### Submisison Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

When both arguments are `int`, the `log(N / (N - n))` term uses integer division for the inner term. This pull request forces floating point division in that term and a couple other places.

#### Intended Effect:

Fixes a bug where int values are different than double values.

#### How to Verify:

Run tests.

#### Side Effects:

None. 

#### Documentation:

None necessary.

#### Reviewer Suggestions: 

Anyone.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by bgoodri** [comment]
Are we sure that adding 0.0 won't get compiled away?

**Comment by bob-carpenter** [comment]
The addition may be compiled away, but the type promotion has to happen because the return type of `operator+(int, double)` is `double`.

Nevertheless, adding 0.0 is a clunky way to cast;  I'd much rather see `static_cast<double>(a)` than `a + 0.0` because it clearly indicates the intent of the operation (and thus providing a clue to any human optimizer who might be tempted to remove `+ 0.0`).

**Comment by syclik** [comment]
The reason I didn't go directly to `static_cast<double>(a)` is because of the generic types. We'd have to do something like: `static_cast<boost::promote<T, double>::type>(a)`.

Now that I look at it, that doesn't look so bad so I'll implement it that way. Unless someone has a better option.

**Comment by bob-carpenter** [comment]
I didn't even catch that those were templated types.  The arguments should be constant references, otherwise, there'll be redundant copies.

The only case where there's an issue is the integer, integer case.  But it seems extreme to write a specialization for that.

If you replace `log(N/(N - n + 0.0))` with `log(N) - log(N - n)` then everything should be OK.  It's an extra call to log, but that can be rolled into the earlier `log(N - n)` call.  But then that needs the promoted type for a local variable, so it's not going to be pretty either way.

Also, the lower expression still has a call to `lgamma(n + 1)`, but `n` could be very large here, so if the `lgamma` calls are problematic, I don't see how this second route is much safer other than when `n` is small.

**Comment by syclik** [comment]
@bob-carpenter @bgoodri: sitting on the pull request for a little bit let me see the right solution. I just created a template specialization for when both `N` and `n` are `int`. That's the only time this could pose a problem. Mind taking a look?

**Comment by bob-carpenter** [comment]
That looks like it leaves the integer arithmetic in place.  Why not just `static_cast<double>()` where necessary?  The answer is always floating point, and integers can't be overloaded to autodiff variables, so it seems like that would be totally safe.

**Comment by syclik** [comment]
I added a new template specialization for when both arguments are ints. In all other instantiations, we won't have a problem with `N - n`. In all other cases, either `N` or `n` will be `double` or `var` and the promotion happens automatically.

I think we may have been too clever in our original implementation. I don't think we need the generality. But... I don't think we can get rid of the template arguments for this to compile all the way through the Stan language. I guess I could try. Right now, in `function_signatures.h`, we have this line: `add_binary("binomial_coefficient_log");` So... it can be instantiated with autodiff variables, even though that use doesn't really make sense.

So... we can't just `static_cast<double>`. We could `static_cast<boost::math::tools::promote_args<T_N, T_n>::type>`.

**Comment by bob-carpenter** [comment]
> On Apr 22, 2016, at 11:06 AM, Daniel Lee <notifications@github.com> wrote:
> 
> I added a new template specialization for when both arguments are ints. In all other instantiations, we won't have a problem with N - n. In all other cases, either N or n will be double or var and the promotion happens automatically.
> 
> I think we may have been too clever in our original implementation. I don't think we need the generality. But... I don't think we can get rid of the template arguments for this to compile all the way through the Stan language. I guess I could try. Right now, in function_signatures.h, we have this line: add_binary("binomial_coefficient_log"); So... it can be instantiated with autodiff variables, even though that use doesn't really make sense.

Ah, I forgot this was binomial coefficient and taking
real arguments.  

> So... we can't just static_cast<double>. We could static_cast<boost::math::tools::promote_args<T_N, T_n>::type>.

Nifty idea.  To avoid overpromotion and repeated promotion,
we could assign to temps up front:

  typename promote_args<T_N>::type N_promoted <- N;
  typename promote_args<T_n>::type n_promoted <- n;

I think I'd prefer one more complicated function than a template
specialization, but it's not a strong preference given how simple
the function is.  

- Bob

**Comment by syclik** [comment]
@bob-carpenter, I think this implementation should be better.

File: stan/math/prim/scal/fun/binomial_coefficient_log.hpp [hunk 1]
```diff
@@ -68,12 +68,14 @@ namespace stan {
       if ((N < cutoff) || (N - n < cutoff)) {
         return lgamma(N + 1.0) - lgamma(n + 1.0) - lgamma(N - n + 1.0);
       } else {
-        return n * log(N - n) + (N + 0.5) * log(N/(N-n))
-          + 1/(12*N) - n - 1/(12*(N-n)) - lgamma(n + 1.0);
+        typename return_type<T_N, T_n>::type N_m_n(N - n);
```
Comment (syclik) [inline stan/math/prim/scal/fun/binomial_coefficient_log.hpp]: Yeah, I see it. I just broke up the `log(N - n)` into a separate call and this avoids the need to even do that.

**Comment by syclik** [comment]
Looks great. Tests pass. We're good to go.

**Pull request merged**